### DISCUSSION TITLE
Best practice for storing state in a Plugin

### DISCUSSION THREAD

**Discussion by Arkessler** [bevyengine/bevy#6352] [Q&A] [answered]
Hi all 👋 I'm fairly new to both Bevy and Rust, so I hope this isn't too dumb of a question, but I'm having trouble grokking how to store state data in a Plugin I'm working on. I'm trying to figure out how to keep a long-lived input hashmap in my input handler Plugin so as to support click and drag functionality. I've defined the Plugin as follows and imported it into my main App. 

```
pub struct InputHandlerPlugin;

impl Plugin for InputHandlerPlugin {
    fn build(&self, app: &mut App) {
        app.add_system(change_tile_on_click);
    }
}
```

The above works fine and dandy, but when I change it to 

```
pub struct InputHandlerPlugin {
    mouse_input_map: HashMap<MouseButton, bool>,
}

impl Plugin for InputHandlerPlugin {
    fn build(&self, app: &mut App) {
        self.mouse_input_map = HashMap::new();
        app.add_system(change_tile_on_click);
    }
}
```

I start getting the following warning in VSCode

```
error: expected value, found struct `InputHandlerPlugin`
label: `InputHandlerPlugin` defined here
```

Clicking through various examples, it seems like all Plugins I can find use unit structs. So I'm assuming the above error might have something to do with me adding fields. My question is, if that's true, where should I keep the `mouse_input_map` in such a way as to make it available to my Plugin and its systems? 

I seem to have gotten my desired behavior working by moving the data to a Resource struct (as follows), but I'm pretty unsure of whether that's appropriate usage of Bevy/Resources.

```
#[derive(Default)]
struct InputData {
    mouse_input_map: HashMap<MouseButton, bool>,
}

pub struct InputHandlerPlugin;

impl Plugin for InputHandlerPlugin {
    fn build(&self, app: &mut App) {
        app.insert_resource(InputData::default());
        app.add_system(store_mouse_button_input);
        app.add_system(change_tile_on_click);
    }
}
```

Thanks in advance for any guidance y'all can offer! 😄

**Comment by alice-i-cecile** [comment, marked as answer]
Yes, using resources is the intended pattern here :) Plugin state information is best used just to configure setup.

**Comment by Arkessler** [comment]
Cool, I think that makes sense, it seems like it's actually pretty easy to get the Resource into whichever Systems need it 😄 . Thanks @alice-i-cecile!
